library/bignum_core.c

```diff
@@ -0,0 +1,294 @@
+/*
+ *  Core bignum functions
+ *
+ *  Copyright The Mbed TLS Contributors
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "common.h"
+
+#if defined(MBEDTLS_BIGNUM_C)
+
+#include <string.h>
+
+#include "mbedtls/error.h"
+#include "mbedtls/platform_util.h"
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_printf      printf
+#define mbedtls_calloc      calloc
+#define mbedtls_free        free
+#endif
+
+#include "bignum_core.h"
+
+size_t mbedtls_mpi_core_clz( const mbedtls_mpi_uint a )
+{
+    size_t j;
+    mbedtls_mpi_uint mask = (mbedtls_mpi_uint) 1 << (biL - 1);
+
+    for( j = 0; j < biL; j++ )
+    {
+        if( a & mask ) break;
+
+        mask >>= 1;
+    }
+
+    return( j );
+}
+
+size_t mbedtls_mpi_core_bitlen( const mbedtls_mpi_uint *A, size_t A_limbs )
```

Comment:
for future reference: I thought we'd decided this would be

`size_t mbedtls_mpi_core_bitlen( const mbedtls_mpi_uint *A, size_t limbs )`

since there is only one `limbs` argument. But please let's not have another round of review unless there's something else to do

Reply:
i.e. "_Length parameters are qualified if necessary_" rather than "_Length parameters are qualified if possible_"

Reply:
UPDATE: there are things that definitely do need to be changed, so it would be good to change this too

Reply:
> for future reference: I thought we'd decided this would be

In my comment above it was still when "if necessary". Later in our slack conversations it was already "if possible". We are qualifying buffer lengths if possible, it is best to align the treatment of limbs. It also makes sense from the perspective of cognitive ease.